channels/rdpecam/client/camera_device_enum_main.c

```diff
@@ -0,0 +1,540 @@
+
+	free(hchannel);
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_on_new_channel_connection(IWTSListenerCallback* pListenerCallback,
+                                           IWTSVirtualChannel* pChannel, BYTE* Data,
+                                           BOOL* pbAccept, IWTSVirtualChannelCallback** ppCallback)
+{
+	GENERIC_LISTENER_CALLBACK* hlistener = (GENERIC_LISTENER_CALLBACK*)pListenerCallback;
+
+	if (!hlistener || !hlistener->plugin)
+		return ERROR_INTERNAL_ERROR;
+
+	WLog_DBG(TAG, "entered");
+	GENERIC_CHANNEL_CALLBACK* hchannel =
+	    (GENERIC_CHANNEL_CALLBACK*)calloc(1, sizeof(GENERIC_CHANNEL_CALLBACK));
+
+	if (!hchannel)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	hchannel->iface.OnDataReceived = ecam_on_data_received;
+	hchannel->iface.OnOpen = ecam_on_open;
+	hchannel->iface.OnClose = ecam_on_close;
+	hchannel->plugin = hlistener->plugin;
+	hchannel->channel_mgr = hlistener->channel_mgr;
+	hchannel->channel = pChannel;
+	*ppCallback = (IWTSVirtualChannelCallback*)hchannel;
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_plugin_initialize(IWTSPlugin* pPlugin, IWTSVirtualChannelManager* pChannelMgr)
+{
+	CameraPlugin* ecam = (CameraPlugin*)pPlugin;
+
+	WLog_DBG(TAG, "entered");
+
+	if (!ecam || !pChannelMgr)
+		return ERROR_INVALID_PARAMETER;
+
+	if (ecam->initialized)
+	{
+		WLog_ERR(TAG, "[%s] plugin initialized twice, aborting", RDPECAM_CONTROL_DVC_CHANNEL_NAME);
+		return ERROR_INVALID_DATA;
+	}
+
+	ecam->version = ECAM_PROTO_VERSION;
+
+	ecam->devices = HashTable_New(FALSE);
+	HashTable_SetupForStringData(ecam->devices, FALSE);
+
+	ecam->hlistener =
+	    (GENERIC_LISTENER_CALLBACK*)calloc(1, sizeof(GENERIC_LISTENER_CALLBACK));
+
+	if (!ecam->hlistener)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	ecam->hlistener->iface.OnNewChannelConnection = ecam_on_new_channel_connection;
+	ecam->hlistener->plugin = pPlugin;
+	ecam->hlistener->channel_mgr = pChannelMgr;
+	const UINT rc = pChannelMgr->CreateListener(pChannelMgr, RDPECAM_CONTROL_DVC_CHANNEL_NAME, 0,
+	                                            &ecam->hlistener->iface, &ecam->listener);
+
+	ecam->initialized = (rc == CHANNEL_RC_OK);
+	return rc;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ecam_plugin_terminated(IWTSPlugin* pPlugin)
+{
+	CameraPlugin* ecam = (CameraPlugin*)pPlugin;
+
+	if (!ecam)
+		return ERROR_INVALID_DATA;
+
+	WLog_DBG(TAG, "entered");
+
+	if (ecam->hlistener)
+	{
+		IWTSVirtualChannelManager* mgr = ecam->hlistener->channel_mgr;
+		if (mgr)
+			IFCALL(mgr->DestroyListener, mgr, ecam->listener);
+	}
+
+	free(ecam->hlistener);
+
+	HashTable_Foreach(ecam->devices, (HASH_TABLE_FOREACH_FN)ecam_dev_destroy, ecam);
```

Comment:
You could probably achieve the same automatically by setting a dtor on the wObject value of the hashtable

Reply:
Maybe, but it's still at least 1 line if code :-)

Reply:
true, but please do not cast function pointers like here (better wrap your function with appropriate arguments) so that compilers warn if something changes.

Reply:
Done, per @hardening  suggestion